### ISSUE TITLE
Zapline not workig on CTF data

### ISSUE THREAD

**Issue by LeonardoMelloni** [snesmaeili/PyZapline_plus#34]
Hi guys,
I have been trying to use mne_denoise.zapline.Zapline to remove line noise (50Hz and multiples) from MEG CTF data (275 magnetometers) with sampling frequency at 1200Hz (7 minutes resting state recordings).

In my automated pipeline I will downsample it to 400Hz and filter between 0.1 and 150Hz. 

In an .ipynb I have been testing following the tutorials, before implementing it for batch processing.

### Attempt 1
Zapline standard doesn't automatically remove any noise component.

```
from mne_denoise.zapline import ZapLine
from mne_denoise.viz import plot_psd_comparison

raw = mne.io.read_raw_ctf(data_fpath, preload=True)

line_freq = 50
down_sfreq = 400

raw.resample(down_sfreq)
sfreq=raw.info['sfreq']

est = ZapLine(
    line_freq=line_freq,
    sfreq=sfreq,
    n_remove="auto", 
    n_harmonics=3,  # 50 100 150 Hz (restricted to 3 so there's no risk to have problems at Nyquist -200Hz-)
)
est.fit(raw)
cleaned = est.transform(raw)


print(f"Components removed: {est.n_removed_}")
print(f"Component scores (eigenvalues): {est.eigenvalues_}")
print(f"Harmonics processed: {est.n_harmonics_}")
```

> Components removed: 0
> Component scores (eigenvalues): [9.88999359e-01 9.68951301e-01 7.14728232e-01 6.76753765e-01
>  5.83699080e-01 4.22202798e-01 1.63730711e-01 5.00206326e-03
>  1.18624482e-03 1.48571576e-04 8.64564508e-05 7.59793074e-05
>  4.66650479e-05 3.74553067e-05 3.18800906e-07]
> Harmonics processed: 3

### Attempt 2
Increase n_remove = 5.
Line noise is still present and I'm loosing broadband power (plus, given that I would like to implement a pipeline the most automated as possible, setting the individual value for each subject is not desirable, I'd rather go for the "auto")

NOTE: changing nkeep, nothing changes in the results.

<img width="584" height="455" alt="Image" src="https://github.com/user-attachments/assets/fc89e00e-3cb9-487b-ba6b-da2232ea1d71" />

### Attempt 3
An LLM suggested me to work on the 
```
raw_mag = raw.copy().pick('mag')
raw_mag._data *= 1e13
```

given that the eigenvalues were very little, but this change has no real impact (it lowers a bit the power at 50Hz, but it is mainly due to a general loss in power. Line noise still remains clearly visible)

### Attempt 4
Zapline plus (adaptive=True).

Same issues as with zapline standard

# Proposal
May it be due to the fact that I am using CTF data with gradient compensation?

Anyway I also tried to work on the data after
`raw.apply_gradient_compensation(0) `

Line noise is still there, but mne-denoise remove noise at other frequencies that comes up when compensation is not applied.

Let me know how to share privately some samples with you if you want to test.

Thanks in advance!!!

**Environment:**
Ubuntu
Python 3.12.13
MNE version:  1.13.0.dev48+g893b7841d
mne_denoise 0.0.1

**Comment by snesmaeili** [comment]
Hi @LeonardoMelloni,

I updated #35 after testing against public Brainstorm CTF recordings. There are two separate failure modes here.

1. `n_remove='auto'` did fail on the eigenvalues you pasted. The previous selector used a `mean + 3 * std` outlier rule, so several co-equal large DSS eigenvalues inflated the mean/std together and none crossed the threshold. The updated auto selector adds a log-space knee detector; on your pasted eigenvalues it selects 7 instead of 0.

2. There was also a CTF full-Raw handling problem. A full CTF `Raw` object can contain magnetometers, reference MEG channels, EEG, and misc channels. Fitting one DSS model across those mixed physical units can make the magnetometer line-noise cleaning ineffective. On a public Brainstorm CTF run, the old full-Raw path retained only 3 DSS dimensions and reduced the magnetometer 60 Hz peak by only 1.38 dB. The updated branch now fits the magnetometers as a homogeneous channel set and preserves non-target channels unchanged.

Public CTF validation, using full `Raw` inputs resampled to 400 Hz and cleaned at 60/120/180 Hz with `n_remove='auto'` and `n_harmonics=3`:

| Dataset | Input | n_removed | 60 Hz reduction | 120 Hz reduction | 180 Hz reduction | Max nearby-flank change | Non-target channel change |
| --- | --- | ---: | ---: | ---: | ---: | ---: | ---: |
| `rest_run1` | 301 ch, 272 mags, comp grade 3 | 10 | 10.05 dB | 0.73 dB | 5.48 dB | 0.17 dB | 0.0 |
| `rest_run2` | 301 ch, 272 mags, comp grade 3 | 10 | 9.75 dB | 0.60 dB | 5.30 dB | 0.14 dB | 0.0 |
| `empty_room` | 299 ch, 272 mags, comp grade 0 | 13 | 33.36 dB | 2.63 dB | 16.12 dB | 0.54 dB | 0.0 |



![Public Brainstorm CTF PSD before/after validation](https://gist.githubusercontent.com/snesmaeili/9f808772b388f8efc1d0ac0615272e73/raw/7fc5483cd720345d6d82566ebe4375df0152504f/ctf_zapline_psd_before_after.svg)

The 120 Hz peak was weak in the two resting runs before cleaning, so the small 120 Hz reduction there is expected. The nearby flanks stayed stable, which is the key check against broadband damage.

Could you try the updated branch on your 50 Hz CTF recording with the full `Raw` object?

```python
import mne
from mne_denoise.zapline import ZapLine

raw = mne.io.read_raw_ctf(data_fpath, preload=True, system_clock="ignore")
raw.resample(400)

est = ZapLine(
    line_freq=50.0,
    sfreq=raw.info["sfreq"],
    n_remove="auto",
    n_harmonics=3,
)
cleaned = est.fit_transform(raw)

print("n_removed:", est.n_removed_)
print("top eigenvalues:", est.eigenvalues_[:15])
```

The important check is not only whether `n_removed_` is around 7. We need the before/after PSD to show that the 50/100/150 Hz peaks drop while nearby flanks remain stable. If auto now selects components but the peaks remain, the next thing to inspect is frequency drift or broadening around the nominal line frequency.
